jax/_src/prng.py

```diff
@@ -0,0 +1,475 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from functools import partial
+from typing import Callable, Iterator, NamedTuple
+import warnings
+
+import numpy as np
+
+from jax import lax
+from jax import core
+from jax import numpy as jnp
+from jax import tree_util
+from jax._src.api import jit, vmap
+from jax.config import config
+from jax.lib import xla_bridge
+from jax.lib import xla_client
+from jax.lib import cuda_prng
+from jax.interpreters import batching
+from jax.interpreters import xla
+from jax._src.pprint_util import pp, vcat
+from jax._src.util import prod
+
+
+UINT_DTYPES = {
+    8: jnp.uint8, 16: jnp.uint16, 32: jnp.uint32, 64: jnp.uint64}  # type: ignore[has-type]
+
+# -- PRNG implementation interface --
+
+class PRNGImpl(NamedTuple):
+  key_shape: core.Shape
+  seed: Callable
+  split: Callable
+  random_bits: Callable
+  fold_in: Callable
+
+  def pprint(self):
+    return (pp(self.__class__.__name__) >> pp(':')) + vcat([
+        pp(k) >> pp(' = ') >> pp(v) for k, v in self._asdict().items()
+    ]).indent(2)
+
+
+# -- PRNG key arrays --
+
+def _is_prng_key_data(impl, keys: jnp.ndarray) -> bool:
+  ndim = len(impl.key_shape)
+  try:
+    return (keys.ndim >= 1 and
+            keys.shape[-ndim:] == impl.key_shape and
+            keys.dtype == np.uint32)
+  except AttributeError:
+    return False
+
+@tree_util.register_pytree_node_class
+class PRNGKeyArray:
+  """An array whose elements are PRNG keys.
+
+  This class lifts the definition of a PRNG, provided in the form of a
+  ``PRNGImpl``, into an array-like pytree class. Instances of this
+  class behave like an array whose base elements are keys, hiding the
+  fact that keys are typically arrays (of ``uint32`` dtype) themselves.
+
+  PRNGKeyArrays are also restricted relative to JAX arrays in that
+  they do not expose arithmetic operations. They instead expose
+  wrapper methods around the PRNG implementation functions (``split``,
+  ``random_bits``, ``fold_in``).
+  """
+
+  impl: PRNGImpl
+  keys: jnp.ndarray
+
+  def __init__(self, impl, key_data: jnp.ndarray):
+    # key_data might be a placeholder python `object` or `bool`
+    # instead of a jnp.ndarray due to tree_unflatten
+    if (type(key_data) not in [object, bool] and
+        not _is_prng_key_data(impl, key_data)):
+      raise TypeError(
+          f'Invalid PRNG key data {key_data} for PRNG implementation {impl}')
+    self.impl = impl
+    self.keys = key_data
+
+  def tree_flatten(self):
+    return (self.keys,), self.impl
+
+  @classmethod
+  def tree_unflatten(cls, impl, keys):
+    keys, = keys
+    return cls(impl, keys)
+
+  @property
+  def dtype(self):
+    # TODO(frostig): remove after deprecation window
+    if config.jax_enable_custom_prng:
+      raise AttributeError("'PRNGKeyArray' has no attribute 'dtype'")
```

Comment:
You might have already explained this to me, but why do we want to remove the ability to do `key.dtype`? Seems like in the current implementation this can only be `unit32`, regardless of the `PRNGImpl`.

Reply:
I'm not sure what to return as the `dtype`? If the purpose is to be an array of opaque keys as elements, then `uint32` seems incorrect (that's the dtype of the data buffer representation of an individual key). I figured that `dtype` is one of the ways in which this array type is not a numpy or jax device array. Perhaps this is too cautious, and if so I'd be happy to revisit it.